pyincore/semanticservice.py

```diff
@@ -0,0 +1,178 @@
+            )
+            raise
+        except requests.exceptions.ConnectionError:
+            logger.error(
+                "ConnectionError: Failed to establish a connection with the server. "
+                "This might be due to a refused connection. "
+                "Please check that you are using the right URLs."
+            )
+            raise
+        except requests.exceptions.RequestException:
+            logger.error(
+                "RequestException: There was an exception while trying to handle your request. "
+                "Please go to the end of this message for more specific information about the exception."
+            )
+            raise
+
+    def get_all_semantic_types(
+        self,
+        hyperlink: bool = False,
+        order: str = "asc",
+        skip: int = 0,
+        limit: int = 50,
+        detail: bool = False,
+        save_json: bool = False,
+        json_path: str = None,
+        timeout: Tuple[int, int] = (30, 600),
+        **kwargs,
+    ) -> Union[list, dict]:
+        """Get all semantic types.
+
+        Args:
+            hyperlink (bool): If True, return a list of hyperlinks to the semantic types.
+            order (str): Order of the semantic types. Can be 'asc' or 'desc'.
+            skip (int): Number of semantic types to skip.
+            limit (int): Number of semantic types to return.
+            detail (bool): If True, return detailed information about the semantic types.
+            save_json (bool): If True, save the json response to a file.
+            json_path (str): Path to save the json response.
+            timeout (tuple): Timeout for the request in seconds.
+                            First value is the connection timeout, second value is the read timeout.
+            **kwargs: Arbitrary keyword arguments.
+
+        Returns:
+            list: List of semantic types.
+
+        """
+
+        url = self.base_url
+        payload = {
+            "hyperlink": hyperlink,
+            "order": order,
+            "skip": skip,
+            "limit": limit,
+            "detail": detail,
+        }
+        response = self.client.get(
+            url, params=payload, timeout=timeout, **kwargs
+        )
+        data = self.return_http_response(response).json()
+
+        if save_json:
+            # if json_path is None: then save to the default location (incore cache folder)
+            if json_path is None:
+                json_path = os.path.join(
+                    pyglobals.PYINCORE_USER_CACHE,
+                    f"semantic_types_ord_{order}_sk_{skip}_lim_{limit}.json",
+                )
+            with open(json_path, "w") as f:
+                json.dump(data, f, indent=4)
+
+        return data
+
+    def get_semantic_type_by_name(
+        self,
+        type_name: str,
+        save_json: bool = False,
+        json_path: str = None,
+        timeout: Tuple[int, int] = (30, 600),
+        **kwargs,
+    ) -> Union[list, dict]:
```

Comment:
why the return is type list and dict? I thought it should be uniuqe?

Reply:
the return type could be either a list or a dict, which is why I had put a union but it seems like even if there is 1 response, it returns a list. I'll change it to just be a list

Reply:
have you looked why this sometimes return list (on the controller)? Shouldn't "name" be unique that it's always just gonna be one?